[WebThing] Initial contribution (replaced by #9555)

grro opened the pull request:
This merge request includes a WebThing binding for openHAB. 
	
The WebThing binding supports an interface to remote devices implementing the Web Thing API.  The Web Thing API is specified by the W3C Consortium (draft). The specification describes how a device or thing can be discovered and linked based on web technologies. 
Essentially, a Web Thing API is a RESTful interface including WebSockets support to provide real-time notifications of events as soon as they happen. 

Please refer to the readme.md for more detailed information

Skinah commented:
Your PR is making changes to the netatmo binding, is that intentional?
Channels in the thing-types.xml are starting with capitals instead of camelCase.
Readme does not list the channels nor explain what they do. 

I am not a reviewer, but when someone comes to look at your binding they usually start at the readme as they probably have never used your binding. It is a good idea to have it written in a way to help the reviewer and new users work out what your binding does.

Hope this is helpful and thanks for contributing.

grro commented:
> Your PR is making changes to the netatmo binding, is that intentional?
> Channels in the thing-types.xml are starting with capitals instead of camelCase.
> Readme does not list the channels nor explain what they do.
> 
> I am not a reviewer, but when someone comes to look at your binding they usually start at the readme as they probably have never used your binding. It is a good idea to have it written in a way to help the reviewer and new users work out what your binding does.
> 
> Hope this is helpful and thanks for contributing.

Thank you for your review and input. The changes to the netatmo binding were by accident. I have reverted this. Furthermore, I fixed the channel-related stuff and I have updated the readme to be hopefully better understandable for less-technical users.

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,273 @@
+ * @author Gregor Roth - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = DiscoveryService.class, configurationPid = "webthingdiscovery.mdns")
```
This shouldn't be necessary.
```suggestion
@Component(service = DiscoveryService.class, configurationPid = "webthingdiscovery.mdns")
```

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,273 @@
+ */
+@NonNullByDefault
+@Component(immediate = true, service = DiscoveryService.class, configurationPid = "webthingdiscovery.mdns")
+public class WebthingDiscoveryService extends AbstractDiscoveryService implements ServiceListener {
```
Is there a reason why you don't use `MDNSDiscoveryParticipant`?

grro replied:
MDNSDiscoveryParticipant expected that only **one** service is behind a detected ServiceInfo. In the case of WebThing, there could be more than a single WebThing resource behind a ServiceInfo (-> host + port). For this reason, MDNSDiscoveryParticipant  works only to detect a single WebThing

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,273 @@
+        // create discovery task for each detected service and process these in parallel to increase total discovery
+        // speed
+        var discoveryTasks = Arrays.stream(serviceInfos).map(DiscoveryTask::new).collect(Collectors.toList());
+        ExecutorService exec = Executors.newCachedThreadPool();
```
You could use the ExecutorService provided by this class (`scheduler`).

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,273 @@
+
+            var id = (uri.getHost() + uri.getPort() + uri.getPath()).replaceAll("\\W", "");
+            var thingUID = new ThingUID(THING_TYPE_UID, id);
+            return Optional.of(DiscoveryResultBuilder.create(thingUID).withThingType(THING_TYPE_UID)
+                    .withProperty("webThingURI", uri).withLabel(description.title).build());
```
Can you add the representation property?

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/ChannelToPropertyLink.java:
```diff
@@ -0,0 +1,77 @@
+            this.propertyName = propertyName;
+            channelHandler.observeChannel(channel.getUID(), this);
+        } else {
+            throw new RuntimeException("property " + propertyName + " does not exits");
```
See above. RuntimeException vs custom exception.

grro replied:
unchecked exceptions replaced by checked onces

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/TypeConverters.java:
```diff
@@ -0,0 +1,181 @@
+            }
+            Color rgb = Color.decode(value);
+            float[] hsb = new float[3];
+            hsb = Color.RGBtoHSB(rgb.getRed(), rgb.getGreen(), rgb.getBlue(), hsb);
+            String hsbString = (hsb[0] * 360) + "," + (hsb[1] * 100) + "," + (hsb[2] * 100);
+            return HSBType.valueOf(hsbString);
```
Is there a reason why you don't use `HSBType.fromRGB()`?

grro replied:
Thanks for the hint. fixed

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/TypeConverters.java:
```diff
@@ -0,0 +1,181 @@
+                textValue = ((Collection) propertyValue).stream()
+                        .reduce("", (entry1, entry2) -> entry1.toString() + "\n" + entry2.toString()).toString();
+            }
+            return StringType.valueOf(textValue.toString());
```
```suggestion
            return StringType.valueOf(textValue);
```

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,9 @@
+
+	<name>WebThing Binding</name>
+	<description>The WebThing binding supports an interface to remote devices implementing the Web Thing API.</description>
+	<author>Gregor Roth</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

grro replied:
OK

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/i18n/webthing_xx_XX.properties:
```diff
@@ -0,0 +1,6 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
+# FIXME: please do not add the file to the repo if you add or change no content
+# binding
+binding.webthing.name = <Your localized Binding name>
+binding.webthing.description = <Your localized Binding description>
+
```
Can this be removed?

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,59 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
Can this be removed?

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,59 @@
+
+	<!-- Sample Thing Type -->
+	<thing-type id="generic">
+		<label>WebThing Binding Thing</label>
```
```suggestion
		<label>Generic WebThing</label>
```

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,59 @@
+	<!-- Sample Thing Type -->
+	<thing-type id="generic">
+		<label>WebThing Binding Thing</label>
+		<description>Sample thing for WebThing Binding</description>
```
The word "Binding" is redundant and can therefore be removed. "Sample" because the description is overwritten dynamically?

grro replied:
fixed.

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,59 @@
+
+		<config-description>
+			<parameter name="webThingURI" type="text" required="true">
+				<label>webThingURI</label>
```
Labels should be human readable.

Guideline is 2-3 words with up to 25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,59 @@
+		<description>Sample thing for WebThing Binding</description>
+
+		<config-description>
+			<parameter name="webThingURI" type="text" required="true">
```
You could specify `<context>url</context>` to get a free validation.

grro replied:
thanks for the remark. I have fixed this

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,59 @@
+		<config-description>
+			<parameter name="webThingURI" type="text" required="true">
+				<label>webThingURI</label>
+				<description>This is the url of the webthing </description>
```
Can you fix the capitalization?

grro replied:
Do you mean the description?. Camel case is OK for parameter name and label, or iisn't it?

fwolter replied:
Yes, the description. But the label should be human readable.

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,59 @@
+	</thing-type>
+
+
+	<channel-type id="number">
+		<item-type>Number</item-type>
+		<label>Webthing Binding Channel</label>
+		<description>Number channel for Webthing Binding</description>
+	</channel-type>
```
I think you don't need to define the Channel types here if they are created fully dynamically.

grro replied:
I have tried this without the channel description. Defining items referring to the auto-created channels worked.

Unfortunately, the oh3 user interface does not detect/show the channels. Should this work? Is this a bug in the new oh3 user interface? I would prefer to remove this channel definition. It feels like a workaround

fwolter replied:
> 
> 
> I have tried this without the channel description. Defining items referring to the auto-created channels worked.

Without the description or without the whole `channel-type` tag? Did you try remove the entire tag?

> Unfortunately, the oh3 user interface does not detect/show the channels. Should this work? Is this a bug in the new oh3 user interface? I would prefer to remove this channel definition. It feels like a workaround

The Channels will only be shown if you add the corresponding `channel` tags in the `thing-type` tag or by adding them dynamically.

grro replied:
I implemented the code to add channel dynamically according to https://www.openhab.org/docs/developer/bindings/#updating-the-thing-structure Unfortunately, the Channels will only be shown without the channel types entries in the thing-types.xml

grro replied:
I have updated my tests to use Junit5

grro replied:
So how do we proceed with topic? Essentially, the current solution works

fwolter replied:
I missed that you need an existing channel type when creating a Channel. It's all fine.

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/PropertyToChannelLink.java:
```diff
@@ -0,0 +1,82 @@
+     * @throws IOException if the WebThing can not be connected
+     */
+    private PropertyToChannelLink(ConsumedThing webThing, String propertyName, ChannelHandler channelHandler,
+            Channel channel) throws IOException {
```
This exception is never thrown.

grro replied:
I have fixed this

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/TypeConverters.java:
```diff
@@ -0,0 +1,181 @@
+        public Command toStateCommand(Object propertyValue) {
+            String textValue = propertyValue.toString();
+            if (propertyValue instanceof Collection) {
+                textValue = ((Collection) propertyValue).stream()
+                        .reduce("", (entry1, entry2) -> entry1.toString() + "\n" + entry2.toString()).toString();
```
Can you fix the compiler warning?

grro replied:
compiler warning will be suppressed

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/TypeMapping.java:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * The item type description
+     */
+    @NonNullByDefault
```
Inner classes don't need to be annotated.

grro replied:
removed

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

There are some compiler warnings about null values left.

Your code doesn't compile as there are checkstyle errors. You could take a look at `target/code-analysis/report.html`.

grro commented on .gitignore:
```diff
@@ -24,3 +24,4 @@ features/**/src/main/feature
 .vscode
 .factorypath
 pom.xml.versionsBackup
+/.github/workflows
```
Sorry, this was done by accident. I have reverted the change

grro commented on bundles/org.openhab.binding.webthing/README.md:
```diff
@@ -0,0 +1,94 @@
+
+| Thing | channel  | type   | description                  |
+|--------|----------|--------|------------------------------|
+| WebThing | Automatic | Automatic | All channels will be generated automatically based on the detected WebThing properties |
```
Channels are completely individual.

grro commented on bundles/org.openhab.binding.webthing/README.md:
```diff
@@ -0,0 +1,94 @@
+
+![discovery picture](docs/speedmonitor.png) 
+
+Alternatively, you may add the WebThing as a openHAB Thing by using a webthing.thing file that has to be located inside the things folder.  
```
The screenshots are screenshots of the OH3#milestone3 user interface. Anyway I have updated the text by referring to the openHAB administration interface

grro commented:
> our code doesn't compile as there are checkstyle errors. You could take a look at `target/code-analysis/report.html`.

I have fixed the checkstyle errors. Thanks for feedback

grro commented on bundles/org.openhab.binding.webthing/pom.xml:
```diff
@@ -0,0 +1,31 @@
+  <name>openHAB Add-ons :: Bundles :: Webthing Binding</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.13</version>
+      <scope>test</scope>
+    </dependency>
```
I have re-added the junit dependency. In my local environment  (mvn on command line/windows) it does not work without defining this dependency

fwolter replied:
I overlooked that you are referencing JUnit 4. We are using JUnit 5. So, your code needs some small adaptions.

grro replied:
I have updated my tests to use Junit5

fwolter commented:
Just to avoid confusion, did you click on "Load more ..." on my review feedback? Because you solved all visible feedback and none of the hidden.

grro commented:
> Just to avoid confusion, did you click on "Load more ..." on my review feedback? Because you solved all visible feedback and none of the hidden.

 Thanks. I have overlooked that. I am going to fix the open issues in the next days

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+    private final Map<ChannelUID, ItemChangedListener> itemChangedListenerMap = new ConcurrentHashMap<>();
+    private final AtomicReference<Optional<ConsumedThing>> webThingConnectionRef = new AtomicReference<>(
+            Optional.empty());
+    private final AtomicReference<Instant> lastReconnect = new AtomicReference<>(Instant.now());
```
webThingConnectionRef, lastReconnect  as well as itemChangedListenerMap will be refreshed if the WebThing connection has been established within the tryReconnect connect. This method will also be called within the watchdog task

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+            var optionalWebThingURI = getWebThingURI();
+            if (optionalWebThingURI.isPresent()) {
+
+                logger.info("try to connect WebThing {}", optionalWebThingURI.get());
```
fixed

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+                logger.info("try to connect WebThing {}", optionalWebThingURI.get());
+                var connected = tryReconnect(optionalWebThingURI.get());
+                if (connected) {
+                    logger.info("WebThing {} connected", getWebThingLabel());
```
fixed

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+                aliveWatchdog.start();
+                aliveWatchdogRef.getAndSet(Optional.of(aliveWatchdog)).ifPresent(AliveWatchdog::close);
+            } else {
+                logger.warn("could not initialize WebThing. URI is not set or invalid. {}",
+                        getConfigAs(WebThingConfiguration.class).webThingURI);
```
fixed

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,220 @@
+     * @throws IOException it the WebThing can not be connected
+     */
+    ConsumedThingImpl(URI webThingURI, Consumer<String> errorHandler) throws IOException {
+        this(webThingURI, errorHandler, HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build(),
+                WebSocketConnectionFactory.instance());
```
is there a specific reason not to use the Java 11 built-in client?

fwolter replied:
If you use the framework's client, the resources/threads are re-used.

grro replied:
done

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+        var webThingUriString = webThingConfiguration.webThingURI;
+        if (webThingUriString == null) {
+            logger.warn("webThing uri has not been set");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "webThing uri has not been set");
```
fixed

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+
+        // and update the thing
+        updateThing(thing);
+        logger.debug("thing updated: {}", print(thing));
```
I have removed the verbose print output

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+    //
+    /////////////
+
+    private final class AliveWatchdog extends Thread {
```
fixed by using scheduler.scheduleWithFixedDelay()

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+                                logger.debug(
+                                        "could not reconnect WebThing {} (periodical reconnect). Next trial in {} sec",
+                                        getWebThingLabel(), HEALTH_CHECK_PERIOD.getSeconds());
+                            }
```
done

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandlerFactory.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        return new WebThingHandler(thing);
```
removed

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+        // create a channel for each WebThing property
+        for (var entry : webThing.getThingDescription().properties.entrySet()) {
+            var channel = Channels.createChannel(thing.getUID(), entry.getKey(), entry.getValue());
+            thingBuilder.withoutChannel(channel.getUID()).withChannel(channel);
```
Overriding an existing channel fails (if not removed before). For this reason, a possible exiting definition will be removed first. The WebThing binding renews the WenThing connection from time to time. In this case (or by an updated WebThing) interface the thingStructureChanged will be called again.

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+                        }
+                    }
+                } catch (Exception e) {
+                    logger.warn("error occurred by running watchdog task", e);
```
has been removed

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/dto/Property.java:
```diff
@@ -0,0 +1,44 @@
+ * @author Gregor Roth - Initial contribution
+ */
+@NonNullByDefault
+public class Property {
```
done

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,220 @@
+                onError("WebThing resource " + webThingURI + " disconnected");
+                throw new IOException("Got error response: " + response.body());
+            }
+            var properties = new Gson().fromJson(response.body(), Map.class);
```
done

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/DescriptionLoader.java:
```diff
@@ -0,0 +1,79 @@
+        } catch (JsonSyntaxException se) {
+            throw new IOException("resource seems not to be a WebThing. Typo?");
+        } catch (InterruptedException ie) {
+            throw new IOException(ie.getMessage());
```
done

fwolter replied:
Did you push your changes?

grro replied:
fixed now

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+                lastReconnect.set(Instant.now());
+                updateStatus(ThingStatus.ONLINE);
+                return true;
+            } catch (Exception e) {
```
Updated by catching the more concrete type

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/channel/Channels.java:
```diff
@@ -0,0 +1,73 @@
+        if (property.unit == null) {
+            channelBuilder.withLabel(property.title);
+        } else {
+            channelBuilder.withLabel(property.title + " [" + property.unit + "]");
```
OK

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,220 @@
+            }
+        } catch (InterruptedException | IOException e) {
+            onError("WebThing resource " + webThingURI + " disconnected");
+            throw new RuntimeException("could not read " + propertyName + " (" + propertyUri + "). " + e.getMessage());
```
unchecked exceptions replaced by checked onces

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,334 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof State) {
```
done

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java:
```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    private final class ConnectionWatchDog extends Thread {
```
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,23 @@
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-webthing" description="WebThing Binding" version="${project.version}">
+		<feature>openhab-runtime-base</feature>
```
You need to add `<feature>openhab-transport-mdns</feature>`.

grro replied:
done

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,220 @@
+    ConsumedThingImpl(URI webthingUrl, Consumer<String> errorHandler, HttpClient httpClient,
+            WebSocketConnectionFactory webSocketConnectionFactory) throws IOException {
+        this(webthingUrl, errorHandler, httpClient, webSocketConnectionFactory, DEFAULT_PING_PERIOD);
+    }
```
changed to jetty

grro commented:
> Thanks for your contribution! I reviewed your code and here is my feedback.
> 
> There are some compiler warnings about null values left.
> 
> Your code doesn't compile as there are checkstyle errors. You could take a look at `target/code-analysis/report.html`.

Thanks for the review. All issues should be solved now. Please give me feedback if there is more to do

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,246 @@
+                }
+            }
+        }
+        throw new RuntimeException(
+                "WebThing " + webThingURI + " does not support a property uri. WebThing description: " + description);
```
Are you aware that your binding will crash if this exception is raised? If this can only happen due to a bug in your code, I'd make this rather an `IllegalStateException`. If this is expected to happen, it's better to make it a checked (custom) exception and handle it in the calling method. Same for below.

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,246 @@
+                throw new PropertyAccessException("could not read " + propertyName + " (" + propertyUri + ")");
+            }
+            var body = response.getContentAsString();
+            var properties = new Gson().fromJson(body, Map.class);
```
You could make the Gson object a final field to save resources. Same for below.

grro replied:
fixed

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,246 @@
+            }
+            var body = response.getContentAsString();
+            var properties = new Gson().fromJson(body, Map.class);
+            var value = properties.get(propertyName);
```
`properties` can be null. You could do a null check to prevent the compiler warning.

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/DescriptionLoader.java:
```diff
@@ -0,0 +1,92 @@
+                        "could not read resource description " + webthingURI + ". Got " + response.getStatus());
+            }
+            var body = response.getContentAsString();
+            var description = new Gson().fromJson(body, WebThingDescription.class);
```
See above. Store Gson to a field.

grro replied:
fixed. I have overlooked this

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java:
```diff
@@ -0,0 +1,169 @@
+    WebSocketConnectionImpl(ScheduledExecutorService executor, Consumer<String> errorHandler, Duration pingPeriod) {
+        this.errorHandler = errorHandler;
+        this.pingPeriod = pingPeriod;
+        this.pingHandle = executor.scheduleAtFixedRate(this::sendPing, pingPeriod.dividedBy(2).toMillis(),
+                pingPeriod.toMillis(), TimeUnit.MILLISECONDS);
```
A ping doesn't seem to be too timing sensitive. Can this be `scheduleWithFixedDelay`?

grro replied:
yes

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java:
```diff
@@ -0,0 +1,169 @@
+    @Override
+    public void onWebSocketText(String message) {
+        try {
+            var propertyStatus = new Gson().fromJson(message, PropertyStatusMessage.class);
```
See above. Store Gson to a field.

grro replied:
fixed. I have overlooked this

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java:
```diff
@@ -0,0 +1,169 @@
+    private void onError(String message) {
+        try {
+            errorHandler.accept(message);
+        } catch (Exception e) {
```
What type of exception do you expect here? Can you specify the concrete type?

fwolter replied:
Can you comment on my question?

grro replied:
catching exception removed

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java:
```diff
@@ -0,0 +1,169 @@
+        if (session != null) {
+            try {
+                session.getRemote().sendPing(ByteBuffer.wrap(Instant.now().toString().getBytes()));
+            } catch (Exception e) {
```
You should specify the concrete exception you want to catch, since catching `Exception` catches also all runtime exceptions.

grro replied:
I handle IOException only now

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,287 @@
+            for (var future : scheduler.invokeAll(discoveryTasks, 5, TimeUnit.MINUTES)) {
+                future.get(5, TimeUnit.MINUTES);
+            }
+        } catch (Exception e) {
```
What type of exception do you expect here? Can you specify the concrete type?

grro replied:
i have fixed this

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,287 @@
+                future.get(5, TimeUnit.MINUTES);
+            }
+        } catch (Exception e) {
+            logger.warn("error occurred by discovering", e);
```
Are you logging the stack trace by intention?

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,287 @@
+                            discoveryResult.getProperties().get("webThingURI"),
+                            discoveryResult.getProperties().get("id"), discoveryResult.getProperties().get("schemas"));
+                }
+            } catch (Exception e) {
+                logger.warn("error occurred by discovering {}", serviceInfo.getNiceTextString(), e);
```
See above

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/ChannelToPropertyLink.java:
```diff
@@ -0,0 +1,83 @@
+            webThing.writeProperty(propertyName, typeConverter.toPropertyValue((State) stateCommand));
+            logger.debug("property {} updated with {} ({}) ", propertyName, propertyValue, this.propertyType);
+        } catch (PropertyAccessException pae) {
+            logger.warn("could not write WebThing property {} with new channel value", propertyName, pae);
```
Is it sufficient to log only the exception's message, instead of the whole stack trace?

grro replied:
reduced to log message  only

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/link/PropertyToChannelLink.java:
```diff
@@ -0,0 +1,81 @@
+            channelHandler.updateItemState(channel.getUID(), stateCommand);
+            logger.debug("channel {} updated with {} ({})", channel.getUID().getAsString(), value,
+                    channel.getAcceptedItemType());
+        } catch (Exception e) {
+            logger.warn("updating channel {} with {} ({}) failed", channel.getUID().getAsString(),
+                    channel.getAcceptedItemType(), value, e);
```
See above. concrete exception and stack trace.

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,61 @@
+			<parameter name="webThingURI" type="text" required="true">
+				<context>url</context>
+				<label>The URI of the WebThing</label>
+				<description>The URI of the WebThing to be connected. E.g. the URI of a web-connected MotionSensor
+					or a URI of a
+					web-connected Display</description>
```
Can you fix the line break?

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    public void onError(String reason) {
+        if (reason == null) {
```
`reason` cannot be null at this place.

grro replied:
fixed

fwolter reviewed (changes requested):
There are some compiler warnings left concerning null values.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

grro commented on bundles/org.openhab.binding.webthing/pom.xml:
```diff
@@ -0,0 +1,17 @@
+
+  <artifactId>org.openhab.binding.webthing</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: Webthing Binding</name>
```
fixed. thanks for the hint

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,311 @@
+        webThingConnectionRef.getAndSet(Optional.empty()).ifPresent(ConsumedThing::close);
+
+        if (wasConnectedBefore) { // to reduce log messages, just log in case of connection state changed
+            logger.info("WebThing {} disconnected {}. Try reconnect (each {} sec)", getWebThingLabel(), reason.trim(),
+                    HEALTH_CHECK_PERIOD.getSeconds());
```
log level reduced to debug

grro commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,311 @@
+    //
+    /////////////
+
+    private final class WatchdogTask implements Runnable {
```
done

fwolter commented:
Are you finished making all changes?

grro commented:
> Are you finished making all changes?

Yes, thank you  for the quick reaction

fwolter commented on bundles/org.openhab.binding.webthing/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.0.0-SNAPSHOT</version>
```
To make the binding compile, you need to rebase your branch and set the version to the upcoming openHAB version: `3.1.0-SNAPSHOT`.

Here are the commands for rebasing your branch:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+                }
+            } else {
+                logger.warn("could not initialize WebThing. URI is not set or invalid. {}",
+                        getConfigAs(WebThingConfiguration.class).webThingURI);
```
Please use `this.webThingURI` here, as `getConfigAs()` is an expensive operation.

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:
```diff
@@ -0,0 +1,302 @@
+     * @param webThing the WebThing that is used for the new structure
+     */
+    private void thingStructureChanged(ConsumedThing webThing) {
+        var thingBuilder = editThing().withLabel(webThing.getThingDescription().title);
```
Labels are expected to be as short as possible. Guideline is 2-3 words with up to 25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

Is this compatible to common WebThing devices?

grro replied:
THe WebThing rfc https://iot.mozilla.org/wot/#title-member does not limit the length of the title. However, I expect that short titles will be used generally

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:
```diff
@@ -0,0 +1,257 @@
+                }
+            }
+        }
+        throw new IllegalStateException(
+                "WebThing " + webThingURI + " does not support a property uri. WebThing description: " + description);
```
I can vaguely remember I asked that already. Do you handle this exception anywhere? Same for below.

grro replied:
remove this exception by introducing an optional return

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java:
```diff
@@ -0,0 +1,175 @@
+        try {
+            errorHandler.accept(message);
+        } catch (Exception e) {
+            logger.warn("error occurred by performing on disconnect", e);
```
Do you log the stack trace by intention?

grro replied:
handling exception has been removed

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,285 @@
+        // speed
+        var discoveryTasks = Arrays.stream(serviceInfos).map(DiscoveryTask::new).collect(Collectors.toList());
+        try {
+            for (var future : scheduler.invokeAll(discoveryTasks, 5, TimeUnit.MINUTES)) {
+                try {
+                    future.get(5, TimeUnit.MINUTES);
+                } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                    logger.warn("discovering task {} terminated. {}", future, e.getMessage());
+                }
+            }
```
If the user cancels the scan process, you need to cancel all tasks.

It is crucial to handle the `InterruptedException` from `loadWebthingDescription()` to finish the tasks immediately when they are cancelled by `stopScan()`.

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:
```diff
@@ -0,0 +1,285 @@
+
+            var thingUID = new ThingUID(THING_TYPE_UID, id);
+            Map<String, Object> properties = new HashMap<>(1);
+            properties.put("id", id);
+            properties.put("schemas", description.contextKeyword);
+            return Optional.of(DiscoveryResultBuilder.create(thingUID).withThingType(THING_TYPE_UID)
+                    .withProperty("webThingURI", uri).withLabel(description.title).withProperties(properties)
+                    .withRepresentationProperty("id").build());
```
Please make the strings constants.

grro replied:
done

fwolter commented on bundles/org.openhab.binding.webthing/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,60 @@
+		<config-description>
+			<parameter name="webThingURI" type="text" required="true">
+				<context>url</context>
+				<label>The URI of the WebThing</label>
```
```suggestion
				<label>URI</label>
```

grro commented:
Unfortunately rebasing causes a lot of trouble regarding the git project. I am not able to fix the 'signed off' issue even though following the help description. Furthermore, my PR includes changes of other bindings which I am not able to remove. How should we go on? Should I open a new PR?

fwolter commented:
The Git history seems very garbled. I think it's easier to open a new PR. Did this happen during fixing the DCO or during rebasing to 3.1.0-SNAPSHOT?

grro commented:
> The Git history seems very garbled. I think it's easier to open a new PR. Did this happen during fixing the DCO or during rebasing to 3.1.0-SNAPSHOT?

Rebasing causes problems.

grro commented:
I have created a new PR https://github.com/openhab/openhab-addons/pull/9555 replacing this one.

Confectrician commented:
You should be able to close this yourself as starter of the pull request.

![](https://docs.github.com/assets/images/help/pull_requests/pullrequest-closebutton.png)

grro commented:
Replaced by https://github.com/openhab/openhab-addons/pull/9555

fwolter commented:
Do you remeber what exactly caused the problems during rebasing, so that I can improve my small rebasing guidance?

grro commented:
> Do you remeber what exactly caused the problems during rebasing, so that I can improve my small rebasing guidance?

I suppose it was my fault. It started with my mistake by including artifacts of other bindings in my initial pull request. To fixed (revert) this I used GitHub Desktop (win os; for ordinary commits, I am using the command line). With rebase the artifacts of the other bindings re-occur.
